server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1158,117 +1254,59 @@ public static List<String> resolve(Context context, List<String> expressions) {
                 return resolvedExpressions;
             }
 
-            Set<String> result = innerResolve(context, expressions, options, metadata);
-
-            if (result == null) {
-                return expressions;
-            }
-            if (result.isEmpty() && options.allowNoIndices() == false) {
-                IndexNotFoundException infe = new IndexNotFoundException((String) null);
-                infe.setResources("index_or_alias", expressions.toArray(new String[0]));
-                throw infe;
+            ExplicitExpressionResolver.Resolved explicitsAndWildcards = ExplicitExpressionResolver.resolve(
+                context,
+                expressions,
+                options,
+                metadata
+            );
+            if (explicitsAndWildcards.wildcardsAndExclusions().isEmpty()) {
+                return new ArrayList<>(explicitsAndWildcards.explicitNames());
+            } else {
+                Set<String> expandedWildcards = innerResolve(context, explicitsAndWildcards.wildcardsAndExclusions(), options, metadata);
```

Comment:
If I understand this correctly, we will always include explicitly mentioned names. Meaning, exclusions will never be evaluated agains explicit names.

I've run a simple test with the following example:
Given indices: `testXXX`, `testXYY` and `testYYY`
and expressions: `testYYY`, `testX*`, `-*`
before refactoring it would result in an `empty set` but after refactoring it returns `testYYY`.

Reply:
If it helps, this is the test I wrote in `WildcardExpressionResolverTests`:

```java
    public void testConvertWildcardsWithExclusion() {
        Metadata.Builder mdBuilder = Metadata.builder()
            .put(indexBuilder("testXXX"))
            .put(indexBuilder("testXYY"))
            .put(indexBuilder("testYYY"));
        ClusterState state = ClusterState.builder(new ClusterName("_name")).metadata(mdBuilder).build();

        IndexNameExpressionResolver.Context context = new IndexNameExpressionResolver.Context(
            state,
            IndicesOptions.lenientExpandOpen(),
            SystemIndexAccessLevel.NONE
        );

        assertThat(
            newHashSet(IndexNameExpressionResolver.WildcardExpressionResolver.resolve(context, Arrays.asList("testYYY", "testX*", "-*"))),
            equalTo(newHashSet())
        );
    }
```

Reply:
🤦 this is a problem, thanks for catching it!
I have pushed [7545d15](https://github.com/elastic/elasticsearch/pull/88678/commits/7545d1559d32a080aee426b72e15356f1ae044a8) .

Reply:
It's a sign we need more tests, I will see to it.

Reply:
I now realize that this code in "core" permits exclusions of names by a different name.
For example, given two indices `index1` and `index2` and an alias, `alias` that refers to both, the expression `index1*,-alias1*` resolves to no indices when Security is disabled, but resolve to `index1` when Security is enabled.
We need to discuss what behavior we want to keep...
I'll organize an email/meeting.